{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Search_engine_data.ipynb",
      "provenance": [],
      "collapsed_sections": [
        "MhN0gWKiBGiQ"
      ],
      "toc_visible": true,
      "authorship_tag": "ABX9TyPhA3UJ+NMCCqgWtPIxD37A",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/AlexanderCoudijzer/IR_data_test/blob/main/Search_engine_data.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Data loading testing\n",
        "see https://ir-datasets.com/beir.html#beir/climate-fever  \n",
        "and https://ir-datasets.com/python.html"
      ],
      "metadata": {
        "id": "XeS4WON90flx"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5Qh8yMNt6QeI"
      },
      "outputs": [],
      "source": [
        "%%capture\n",
        "!pip install --upgrade ir_datasets\n",
        "\n",
        "import pandas as pd\n",
        "from sklearn.feature_extraction.text import CountVectorizer\n",
        "import numpy as np\n",
        "import ir_datasets"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Climate-fever test"
      ],
      "metadata": {
        "id": "MhN0gWKiBGiQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dataset = ir_datasets.load(\"beir/climate-fever\")"
      ],
      "metadata": {
        "id": "-FEjRFse6WS-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dataset.has_docs(), dataset.has_queries(), dataset.has_qrels(), dataset.has_scoreddocs()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "H9iB6Lpj9mqH",
        "outputId": "03dc86e0-9c3d-4b8e-c932-c434840d92de"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(True, True, True, False)"
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%capture\n",
        "for doc in dataset.docs_iter()[:1]:\n",
        "    print(doc)"
      ],
      "metadata": {
        "id": "QvkZ-HaN-zLx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dataset.docs_cls().__annotations__"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vZ6D8j-LiZeO",
        "outputId": "f46a245d-afa0-4c5e-8f82-e473e5b219cc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "OrderedDict([('doc_id', str), ('text', str), ('title', str)])"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "for doc in dataset.docs_iter()[1000:1010]:\n",
        "    print(doc)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "U5dnvRfG2M2K",
        "outputId": "93550da3-4718-45ae-dbd7-0152b4ecc47c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "BeirTitleDoc(doc_id='179th_Fighter_Squadron', text='The 179th Fighter Squadron ( 179 FS ) is a unit of the Minnesota Air National Guard 148th Fighter Wing located at Duluth Air National Guard Base , Minnesota . The 179th is equipped with the General Dynamics F-16C Fighting Falcon .', title='179th Fighter Squadron')\n",
            "BeirTitleDoc(doc_id='1922_VFL_Grand_Final', text=\"The 1922 VFL Grand Final was an Australian rules football game contested between the Fitzroy Football Club and Collingwood Football Club , held at the Melbourne Cricket Ground in Melbourne on 21 October 1922 . It was the 26th annual Grand Final of the Victorian Football League , staged to determine the premiers for the 1922 VFL season .   The half-time break was more than thirty minutes .   The match , attended by 50,064 spectators , was won by Fitzroy by a margin of 11 points , marking that club 's seventh premiership victory .\", title='1922 VFL Grand Final')\n",
            "BeirTitleDoc(doc_id='11th_Infantry_Regiment_(Greece)', text=\"The 11th Infantry Regiment ( 11ο Σύνταγμα Πεζικού '' Αργυρόκαστρο '' , 11ο ΣΠ ) is an infantry regiment of the Hellenic Army . It was founded in March 1897 in Larissa as part of the 2nd Infantry Division 's 4th Brigade . With the Army reorganization of 1912 , it was subordinated to the 4th Infantry Division , where it has remained since . The regiment distinguished itself in the Asia Minor Campaign , earning the highest Greek decoration for valour , the Commander 's Cross of the Cross of Valour . The regiment was disbanded on 2 May 1941 , following the Greek army 's capitulation after the German invasion of Greece , and was reconstituted in 1945 as a recruit training center . In 1965 it was moved to Tripoli , which has been its base since .\", title='11th Infantry Regiment (Greece)')\n",
            "BeirTitleDoc(doc_id='1973_in_Cambodia', text='The following lists events that happened during 1973 in Cambodia .', title='1973 in Cambodia')\n",
            "BeirTitleDoc(doc_id='1903_in_Sweden', text='Events from the year 1903 in Sweden', title='1903 in Sweden')\n",
            "BeirTitleDoc(doc_id='1979_in_Jordan', text='Events from the year 1979 in Jordan .', title='1979 in Jordan')\n",
            "BeirTitleDoc(doc_id='1883_Harvard_Crimson_football_team', text='The 1883 Harvard Crimson football team represented Harvard University in the 1883 college football season . They finished with an 8 -- 2 record .', title='1883 Harvard Crimson football team')\n",
            "BeirTitleDoc(doc_id='1920_Rice_Owls_football_team', text='The 1920 Rice Owls football team represented Rice University during the 1920 college football season .', title='1920 Rice Owls football team')\n",
            "BeirTitleDoc(doc_id='1989_Sovran_Bank_Classic_–_Singles', text='Jimmy Connors was the defending champion but did not compete that year .   Tim Mayotte won in the final 3 -- 6 , 6 -- 4 , 7 -- 5 against Brad Gilbert .', title='1989 Sovran Bank Classic – Singles')\n",
            "BeirTitleDoc(doc_id='1998_Singapore_Open_–_Singles', text=\"The 1998 Singapore Open was a men 's tennis tournament played on Indoor Hard in Singapore , Singapore that was part of the Championship Series of the 1998 ATP Tour . It was the seventh edition of the tournament and was held from 12 October -- 18 October .\", title='1998 Singapore Open – Singles')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "dataset.queries_count()\n",
        "for q in dataset.queries_iter():\n",
        "    print(q)"
      ],
      "metadata": {
        "id": "XXFW20orame4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dataset.qrels_cls().__annotations__ \n",
        "dataset.qrels_defs() , dataset.qrels_count()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "36GP9TUIb5VK",
        "outputId": "2bf0175e-7d5e-4051-fadd-bbe08e193bb3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "({}, 4681)"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "rel_doc = []\n",
        "for qrel in dataset.qrels_iter():\n",
        "    rel_doc.append(qrel[1])\n",
        "len(set(rel_doc))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xreM50wMdaKT",
        "outputId": "a35d6487-72a2-483c-d914-c4ff991f0a60"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[INFO] [starting] opening zip file\n",
            "[INFO] [finished] opening zip file [3ms]\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "1344"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "corpus=[]\n",
        "for doc in dataset.docs_iter():\n",
        "    if doc[0] in rel_doc:\n",
        "        corpus.append([doc[0],doc[1],doc[2]])"
      ],
      "metadata": {
        "id": "6GQqcEqZg1iJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "len(corpus)\n",
        "#corpus"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1e6K7y7whOxM",
        "outputId": "b7d2262f-6c4a-4970-df81-d4f880036a67"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "1344"
            ]
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from sklearn.feature_extraction.text import CountVectorizer\n",
        "import numpy as np"
      ],
      "metadata": {
        "id": "AMs6YG5jeIPi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "vectorizer = CountVectorizer(stop_words='english', strip_accents='ascii')\n",
        "\n",
        "documents_vectorized = vectorizer.fit_transform([doc[1] for doc in corpus])\n",
        "vocabulary = vectorizer.get_feature_names_out()\n",
        "vocabulary\n",
        "dataframe = pd.DataFrame(documents_vectorized.toarray(), columns=vocabulary)\n",
        "dataframe"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "P2NcSt6oeJU2",
        "outputId": "2b8b7351-9704-49ed-ffa5-40270e237cf1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      00  000  0005  001  003  005  006  007  01  010  ...  zoogeographical  \\\n",
              "0      0    0     0    0    0    0    0    0   0    0  ...                0   \n",
              "1      0    1     0    0    0    0    0    0   0    0  ...                0   \n",
              "2      0    0     0    0    0    0    0    0   0    0  ...                0   \n",
              "3      0    1     0    0    0    0    0    0   0    0  ...                0   \n",
              "4      0    0     0    0    0    0    0    0   0    0  ...                0   \n",
              "...   ..  ...   ...  ...  ...  ...  ...  ...  ..  ...  ...              ...   \n",
              "1339   0    0     0    0    0    0    0    0   0    0  ...                0   \n",
              "1340   0    0     0    0    0    0    0    0   0    0  ...                0   \n",
              "1341   0    0     0    0    0    0    0    0   0    0  ...                0   \n",
              "1342   0    0     0    0    0    0    0    0   0    0  ...                0   \n",
              "1343   0    2     0    0    0    0    0    0   0    0  ...                0   \n",
              "\n",
              "      zoology  zoon  zoonotic  zooplankton  zoos  zooxanthellae  \\\n",
              "0           0     0         0            0     0              0   \n",
              "1           0     0         0            0     0              0   \n",
              "2           0     0         0            0     0              0   \n",
              "3           0     0         0            0     0              0   \n",
              "4           0     0         0            0     0              0   \n",
              "...       ...   ...       ...          ...   ...            ...   \n",
              "1339        0     0         0            0     0              0   \n",
              "1340        0     0         0            0     0              0   \n",
              "1341        0     0         0            0     0              0   \n",
              "1342        0     0         0            0     0              0   \n",
              "1343        0     0         0            0     0              0   \n",
              "\n",
              "      zooxanthellate  zosteraceae  zvezda  \n",
              "0                  0            0       0  \n",
              "1                  0            0       0  \n",
              "2                  0            0       0  \n",
              "3                  0            0       0  \n",
              "4                  0            0       0  \n",
              "...              ...          ...     ...  \n",
              "1339               0            0       0  \n",
              "1340               0            0       0  \n",
              "1341               0            0       0  \n",
              "1342               0            0       0  \n",
              "1343               0            0       0  \n",
              "\n",
              "[1344 rows x 22638 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-380cfee5-7646-4b45-bb1a-c46e426b32f2\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>00</th>\n",
              "      <th>000</th>\n",
              "      <th>0005</th>\n",
              "      <th>001</th>\n",
              "      <th>003</th>\n",
              "      <th>005</th>\n",
              "      <th>006</th>\n",
              "      <th>007</th>\n",
              "      <th>01</th>\n",
              "      <th>010</th>\n",
              "      <th>...</th>\n",
              "      <th>zoogeographical</th>\n",
              "      <th>zoology</th>\n",
              "      <th>zoon</th>\n",
              "      <th>zoonotic</th>\n",
              "      <th>zooplankton</th>\n",
              "      <th>zoos</th>\n",
              "      <th>zooxanthellae</th>\n",
              "      <th>zooxanthellate</th>\n",
              "      <th>zosteraceae</th>\n",
              "      <th>zvezda</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1339</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1340</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1341</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1342</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1343</th>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1344 rows × 22638 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-380cfee5-7646-4b45-bb1a-c46e426b32f2')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-380cfee5-7646-4b45-bb1a-c46e426b32f2 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-380cfee5-7646-4b45-bb1a-c46e426b32f2');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 21
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "dfs = (dataframe > 0).sum(axis=0)\n",
        "dfs"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ERFU2TBDSQES",
        "outputId": "66bb1a9c-3b2a-4257-84f7-fd89eb7b24e6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "00                  4\n",
              "000               173\n",
              "0005                1\n",
              "001                 1\n",
              "003                 1\n",
              "                 ... \n",
              "zoos                1\n",
              "zooxanthellae       1\n",
              "zooxanthellate      1\n",
              "zosteraceae         1\n",
              "zvezda              1\n",
              "Length: 22638, dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "N = dataframe.shape[0]\n",
        "idfs = np.log10(N/dfs)\n",
        "\n",
        "# Lets first define all the variables we need:\n",
        "k_1 = 1.2 # single value\n",
        "b = 0.8 # single value\n",
        "## considering all words in doc\n",
        "dls = [len(d.split(' ')) for d in [doc[1] for doc in corpus]] # vector\n",
        "## considering words minus stop words in doc. \n",
        "dls = dataframe.sum(axis=1).tolist()\n",
        "avgdl = np.mean(dls) # single value\n",
        "\n",
        "numerator = np.array((k_1 + 1) * dataframe)\n",
        "denominator = np.array(k_1 *((1 - b) + b * (dls / avgdl))).reshape(N,1) + np.array(dataframe)\n",
        "BM25_tf = numerator / denominator\n",
        "idfs = np.array(idfs)\n",
        "BM25_score = BM25_tf * idfs\n",
        "\n",
        "bm25_idf = pd.DataFrame(BM25_score, columns=vocabulary)\n",
        "bm25_idf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "GaLk2DtfYbkh",
        "outputId": "7e61bb4c-c605-4146-d10d-d597d0e492e1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       00       000  0005  001  003  005  006  007   01  010  ...  \\\n",
              "0     0.0  0.000000   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "1     0.0  1.143564   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "2     0.0  0.000000   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "3     0.0  1.102436   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "4     0.0  0.000000   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "...   ...       ...   ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
              "1339  0.0  0.000000   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "1340  0.0  0.000000   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "1341  0.0  0.000000   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "1342  0.0  0.000000   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "1343  0.0  1.297376   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   \n",
              "\n",
              "      zoogeographical  zoology  zoon  zoonotic  zooplankton  zoos  \\\n",
              "0                 0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "1                 0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "2                 0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "3                 0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "4                 0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "...               ...      ...   ...       ...          ...   ...   \n",
              "1339              0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "1340              0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "1341              0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "1342              0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "1343              0.0      0.0   0.0       0.0          0.0   0.0   \n",
              "\n",
              "      zooxanthellae  zooxanthellate  zosteraceae  zvezda  \n",
              "0               0.0             0.0          0.0     0.0  \n",
              "1               0.0             0.0          0.0     0.0  \n",
              "2               0.0             0.0          0.0     0.0  \n",
              "3               0.0             0.0          0.0     0.0  \n",
              "4               0.0             0.0          0.0     0.0  \n",
              "...             ...             ...          ...     ...  \n",
              "1339            0.0             0.0          0.0     0.0  \n",
              "1340            0.0             0.0          0.0     0.0  \n",
              "1341            0.0             0.0          0.0     0.0  \n",
              "1342            0.0             0.0          0.0     0.0  \n",
              "1343            0.0             0.0          0.0     0.0  \n",
              "\n",
              "[1344 rows x 22638 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-291f4fc6-1644-46c9-a961-8cd8e2a94e12\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>00</th>\n",
              "      <th>000</th>\n",
              "      <th>0005</th>\n",
              "      <th>001</th>\n",
              "      <th>003</th>\n",
              "      <th>005</th>\n",
              "      <th>006</th>\n",
              "      <th>007</th>\n",
              "      <th>01</th>\n",
              "      <th>010</th>\n",
              "      <th>...</th>\n",
              "      <th>zoogeographical</th>\n",
              "      <th>zoology</th>\n",
              "      <th>zoon</th>\n",
              "      <th>zoonotic</th>\n",
              "      <th>zooplankton</th>\n",
              "      <th>zoos</th>\n",
              "      <th>zooxanthellae</th>\n",
              "      <th>zooxanthellate</th>\n",
              "      <th>zosteraceae</th>\n",
              "      <th>zvezda</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0.0</td>\n",
              "      <td>1.143564</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0.0</td>\n",
              "      <td>1.102436</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1339</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1340</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1341</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1342</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1343</th>\n",
              "      <td>0.0</td>\n",
              "      <td>1.297376</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1344 rows × 22638 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-291f4fc6-1644-46c9-a961-8cd8e2a94e12')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-291f4fc6-1644-46c9-a961-8cd8e2a94e12 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-291f4fc6-1644-46c9-a961-8cd8e2a94e12');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 24
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "query_num = 23\n",
        "queries = []\n",
        "for q in dataset.queries_iter():\n",
        "    queries.append([q[0],q[1]])\n",
        "print([queries[query_num][1]])\n",
        "\n",
        "vectorizer_q = CountVectorizer(stop_words='english', strip_accents='ascii')\n",
        "query_vectorized = vectorizer_q.fit_transform([queries[query_num][1]])\n",
        "q_terms = vectorizer_q.get_feature_names_out()\n",
        "q_terms"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vPO9WocLc0-C",
        "outputId": "9d577180-5ebb-43d0-ce0c-ea29a743c608"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['Earth’s mean temperature over the last 2,000 years shows two previous periods when temperatures were warmer than they are now; from 1–200 A.D., an epoch called the Roman Warm Period, and more recently the Medieval Warm Period from 900–1100 A.D.']\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['000', '1200', '9001100', 'called', 'earths', 'epoch', 'mean',\n",
              "       'medieval', 'period', 'periods', 'previous', 'recently', 'roman',\n",
              "       'shows', 'temperature', 'temperatures', 'warm', 'warmer', 'years'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 59
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "q_terms = [term for term in q_terms if term in bm25_idf.columns]\n",
        "q_terms_only_df = bm25_idf[q_terms]\n",
        "score_q_d = q_terms_only_df.sum(axis=1)\n",
        "documents = [doc[1] for doc in corpus]\n",
        "sorted(zip(documents,score_q_d.values), key = lambda tup:tup[1], reverse=True)[:10]\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nwR1Lr9YZZ-q",
        "outputId": "4a83c1e4-1467-4249-8dba-ac7487dd2c5b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[(\"For information on the description of the Medieval Warm Period and Little Ice Age in various IPCC reports see MWP and LIA in IPCC reportsThe temperature record of the past 1,000 years is reconstructed using data from climate proxy records in conjunction with the modern instrumental temperature record which only covers the last 150 years at a global scale . Large-scale reconstructions covering part or all of the 1st millennium and 2nd millennium have shown that recent temperatures are exceptional : the Intergovernmental Panel on Climate Change Fourth Assessment Report of 2007 concluded that `` Average Northern Hemisphere temperatures during the second half of the 20th century were very likely higher than during any other 50-year period in the last 500 years and likely '' the highest in at least the past 1,300 years . '' The curve shown in graphs of these reconstructions is widely known as the hockey stick graph because of the sharp increase in temperatures during the last century . As of 2010 this broad pattern was supported by more than two dozen reconstructions , using various statistical methods and combinations of proxy records , with variations in how flat the pre-20th-century `` shaft '' appears . Sparseness of proxy records results in considerable uncertainty for earlier periods .   Individual proxy records , such as tree ring widths and densities used in dendroclimatology , are calibrated against the instrumental record for the period of overlap . Networks of such records are used to reconstruct past temperatures for regions : tree ring proxies have been used to reconstruct Northern Hemisphere extratropical temperatures ( within the tropics trees do not form rings ) but are confined to land areas and are scarce in the Southern Hemisphere which is largely ocean . Wider coverage is provided by multiproxy reconstructions , incorporating proxies such as lake sediments , ice cores and corals which are found in different regions , and using statistical methods to relate these sparser proxies to the greater numbers of tree ring records . The `` Composite Plus Scaling '' ( CPS ) method is widely used for large-scale multiproxy reconstructions of hemispheric or global average temperatures ; this is complemented by Climate Field Reconstruction ( CFR ) methods which show how climate patterns have developed over large spatial areas , making the reconstruction useful for investigating natural variability and long-term oscillations as well as for comparisons with patterns produced by climate models .   During the 1,900 years before the 20th century , it is likely that the next warmest period was from 950 to 1100 , with peaks at different times in different regions . This has been called the Medieval Warm Period , and some evidence suggests widespread cooler conditions during a period around the 17th century known as the Little Ice Age . In the hockey stick controversy , contrarians have asserted that the Medieval Warm Period was warmer than at present , and have disputed the data and methods of climate reconstructions .\",\n",
              "  13.223710024475821),\n",
              " ('The Medieval Warm Period ( MWP ) , Medieval Climate Optimum , or Medieval Climatic Anomaly was a time of warm climate in the North Atlantic region that may also have been related to other climate events around the world during that time , including China and other areas , lasting from about 950 to 1250 . It was followed by a cooler period in the North Atlantic and elsewhere termed the Little Ice Age . Some refer to the event as the Medieval Climatic Anomaly as this term emphasizes that effects other than temperature were important .   Despite uncertainties , especially for the period prior to 1600 for which data are scarce , the warmest period of the last 2,000 years prior to the 20th century in the Northern Hemisphere very likely occurred between 950 and 1100 . Proxy records show peak warmth occurred at different times for different regions , indicating that the Medieval Warm Period was not a time of globally uniform change . Temperatures in some regions matched or exceeded recent temperatures in these regions , but globally the Medieval Warm Period was cooler than recent global temperatures .   Possible causes of the Medieval Warm Period include increased solar activity , decreased volcanic activity , and changes to ocean circulation .',\n",
              "  12.556785623612955),\n",
              " ('The global temperature record shows the fluctuations of the temperature of the atmosphere and the oceans through various spans of time.  The most detailed information exists since 1850, when methodical thermometer-based records began.  There are numerous estimates of temperatures since the end of the Pleistocene glaciation, particularly during the current Holocene epoch.  Older time periods are studied by paleoclimatology.',\n",
              "  10.881701805003832),\n",
              " (\"An ice age is a period of long-term reduction in the temperature of Earth 's surface and atmosphere , resulting in the presence or expansion of continental and polar ice sheets and alpine glaciers . Within a long-term ice age , individual pulses of cold climate are termed `` glacial periods '' ( or alternatively `` glacials '' or `` glaciations '' or colloquially as `` ice age '' ) , and intermittent warm periods are called `` interglacials '' . In the terminology of glaciology , ice age implies the presence of extensive ice sheets in both northern and southern hemispheres . By this definition , we are in an interglacial period -- the Holocene -- of the ice age . The ice age began 2.6 million years ago at the start of the Pleistocene epoch , because the Greenland , Arctic , and Antarctic ice sheets still exist .\",\n",
              "  10.876155401738526),\n",
              " (\"A global warming hiatus , also sometimes referred to as a global warming pause or a global warming slowdown , is a period of relatively little change in globally averaged surface temperatures . In the current episode of global warming many such 15-year periods appear in the surface temperature record , along with robust evidence of the long-term warming trend ; climate is classically averaged over 30-year periods .   Publicity has surrounded claims of a global warming hiatus during the period 1998 -- 2013 . The exceptionally warm El Niño year of 1998 was an outlier from the continuing temperature trend , and so subsequent annual temperatures gave the appearance of a hiatus : by January 2006 , it appeared to some that global warming had stopped or paused . A 2009 study showed that decades without warming were not exceptional , and in 2011 a study showed that if allowances were made for known variability , the rising temperature trend continued unabated . There was increased public interest in 2013 in the run-up to publication of the IPCC Fifth Assessment Report , and despite concerns that a 15-year period was too short to determine a meaningful trend , the IPCC included a section on a hiatus , which it defined as a much smaller increasing linear trend over the 15 years from 1998 to 2012 , than over the 60 years from 1951 to 2012 . Various studies examined possible causes of the short-term slowdown . Even though the overall climate system has continued to accumulate energy due to Earth 's positive energy budget , the available temperature readings at the Earth 's surface indicate slower rates of increase in surface warming than in the prior decade . Since measurements at the top of the atmosphere show that Earth is receiving more energy than it is radiating back into space , the retained energy should be producing warming in the Earth 's climate system .   Research reported in July 2015 on an updated NOAA dataset casts doubt on the existence of a hiatus , and it finds no indication of a slowdown even in earlier years . Scientists working on other datasets welcomed this study , though they have expressed the view that the recent warming trend was less than in previous periods of the same length . Subsequently , a detailed study supports the conclusion that warming is continuing , but it also find there was less warming between 2001 and 2010 than climate models had predicted , and that this slowdown might be attributed to short-term variations in the Pacific decadal oscillation ( PDO ) , which was negative during that period . Another review finds `` no substantive evidence '' of a pause in global warming . A statistical study of global temperature data since 1970 concludes that the term ` hiatus ' or ` pause ' is not justified .   Independent of these discussions about data and measurements for earlier years , 2015 turned out to be much warmer than any of the earlier years , already before El Niño conditions started . The warmth of 2015 largely ended any remaining scientific credibility of claims that the supposed `` hiatus '' since 1998 had any significance for the long-term warming trend . In January 2017 , a study published in the journal Science Advances cast further doubt on the existence of a recent pause , with more evidence that ocean temperatures have been underestimated . An April 2017 study found the data consistent with a steady warming trend globally since the 1970s , with fluctuations within the expected range of short term variability .\",\n",
              "  10.023910465104048),\n",
              " (\"The Younger Dryas is a geological period from c. 12,900 to c. 11,700 calendar years ago ( BP ) . It is named after an indicator genus , the alpine-tundra wildflower Dryas octopetala . Leaves of Dryas octopetala are occasionally abundant in the Late Glacial , often minerogenic-rich , like the lake sediments of Scandinavian lakes . The Younger Dryas saw a sharp decline in temperature over most of the northern hemisphere , at the end of the Pleistocene epoch , immediately before the current warmer Holocene . It was the most recent and longest of several interruptions to the gradual warming of the Earth 's climate since the severe Last Glacial Maximum , c. 27,000 to 24,000 calendar years BP . The change was relatively sudden , taking place in decades , and resulted in a decline of 2 to 6 degrees Celsius , advances of glaciers and drier conditions , over much of the temperate northern hemisphere . It is thought to have been caused by a decline in the strength of the Atlantic meridional overturning circulation , which transports warm water from the equator towards the North Pole , and which in turn is thought to have been caused by an influx of fresh cold water from North America into the Atlantic . The Younger Dryas was a period of climatic change , but the effects were complex and variable . In the southern hemisphere , and some areas of the north such as southeastern North America , there was a slight warming .   The presence of a distinct cold period at the end of the Late Glacial interval has been known for a long time . Paleobotanical and lithostratigraphic studies of Swedish and Danish bog and lake sites , e.g. the Allerød clay pit in Denmark , first recognized and described the Younger Dryas .   The Younger Dryas is the youngest and longest of three stadials that resulted from typically abrupt climatic changes that took place over the last 16,000 calendar years . Within the Blytt-Sernander classification of north European climatic phases , the prefix ` Younger ' refers to the recognition that this original ` Dryas ' period was preceded by a warmer stage , the Allerød oscillation , which in turn was preceded by the Older Dryas around 14,000 calendar years BP . This is not securely dated , and estimates vary by 400 years , but it is generally accepted that it lasted around 200 years . In northern Scotland the glaciers were thicker and more extensive than during the Younger Dryas . The Older Dryas , in turn , is preceded by another warmer stage , the Bølling oscillation that separates it from a third and even older stadial . This stadial is often , but not always , known as the Oldest Dryas . The Oldest Dryas occurred approximately 1,770 calendar years before the Younger Dryas and lasted about 400 calendar years . According to the GISP2 ice core from Greenland , the Oldest Dryas occurred between about 15,070 and 14,670 calendar years BP .   In Ireland , the Younger Dryas has also been known as the Nahanagan Stadial , while in Great Britain it has been called the Loch Lomond Stadial . In the Greenland Summit ice core chronology , the Younger Dryas corresponds to Greenland Stadial 1 ( GS-1 ) . The preceding Allerød warm period ( interstadial ) is subdivided into three events : Greenland Interstadial-1c to 1a ( GI-1c to GI-1a ) .\",\n",
              "  9.958737686600335),\n",
              " (\"The Eemian ( also called the last interglacial , Sangamonian , Ipswichian , Mikulin , Kaydaky , Valdivia or Riss-Würm ) was the interglacial period which began about 130,000 years ago and ended about 115,000 years ago . It corresponds to Marine Isotope Stage 5e . Although sometimes referred to as the `` last interglacial '' ( in the `` most recent previous '' sense of `` last '' ) , it was the second-to-latest interglacial period of the current Ice Age , the most recent being the Holocene which extends to the present day ( having followed the last glacial period ) . The prevailing Eemian climate is believed to have been warmer than that of the Holocene .   The Eemian is known as the Ipswichian in the UK , the Mikulin interglacial in Russia , the Valdivia interglacial in Chile and the Riss-Würm interglacial in the Alps . Depending on how a specific publication defines the Sangamonian Stage of North America , the Eemian is equivalent to either all or part of it .\",\n",
              "  9.318022034500585),\n",
              " ('The Last Glacial Period (LGP) occurred from the end of the Eemian to the end of the Younger Dryas, encompassing the period c.\\u2009115,000 – c.\\u200911,700 years ago. The LGP is part of a larger sequence of glacial and interglacial periods known as the Quaternary glaciation which started around 2,588,000 years ago and is ongoing.  The definition of the Quaternary as beginning 2.58 million years ago is based on the formation of the Arctic ice cap. The Antarctic ice sheet began to form earlier, at about 34 Ma, in the mid-Cenozoic (Eocene–Oligocene extinction event). The term Late Cenozoic Ice Age is used to include this early phase.During this last glacial period there were alternating episodes of glacier advance and retreat. Within the last glacial period the Last Glacial Maximum was approximately 22,000 years ago. While the general pattern of global cooling and glacier advance was similar, local differences in the development of glacier advance and retreat make it difficult to compare the details from continent to continent (see picture of ice core data below for differences). Approximately 12,800 years ago, the Younger Dryas, the most recent glacial epoch, began, a coda to the preceding 100,000 year glacial period. Its end about 11,550 years ago marked the beginning of the Holocene, the current geological epoch. From the point of view of human archaeology, the last glacial period falls in the Paleolithic and early Mesolithic periods. When the glaciation event started, Homo sapiens were confined to lower latitudes and used tools comparable to those used by Neanderthals in western and central Eurasia and by Denisovans and Homo erectus in Asia. Near the end of the event, Homo sapiens migrated into Eurasia and Australia. Archaeological and genetic data suggest that the source populations of Paleolithic humans survived the last glacial period in sparsely wooded areas and dispersed through areas of high primary productivity while avoiding dense forest cover.',\n",
              "  8.434431277047016),\n",
              " (\"The Paleocene -- Eocene Thermal Maximum ( PETM ) , alternatively ( ETM1 ) , and formerly known as the `` Initial Eocene '' or '' '' was a time period with more than 8 ° C warmer global average temperature than today . This climate event began at the time boundary between the Paleocene and Eocene geological epochs . The exact age and duration of the event is uncertain but it is estimated to have occurred around 55.5 million years ago .   The associated period of massive carbon injection into the atmosphere has been estimated to have lasted no longer than 20,000 years . The entire warm period lasted for about 200,000 years . Global temperatures increased by 5 -- 8 ° C . The carbon dioxide was likely released in two pulses , the first lasting less than 2,000 years . Such a repeated carbon release is in line with current global warming . A main difference is that during the Paleocene -- Eocene Thermal Maximum , the planet was essentially ice-free .   The onset of the Paleocene -- Eocene Thermal Maximum has been linked to an initial 5 ° C temperature rise and to extreme changes in Earth 's carbon cycle . The period is marked by a prominent negative excursion in carbon stable isotope records from around the globe ; more specifically , there was a large decrease in 13C/12C ratio of marine and terrestrial carbonates and organic carbon .   Stratigraphic sections of rock from this period reveal numerous other changes . Fossil records for many organisms show major turnovers . For example , in the marine realm , a mass extinction of benthic foraminifera , a global expansion of subtropical dinoflagellates , and an appearance of excursion , planktic foraminifera and calcareous nanofossils all occurred during the beginning stages of PETM . On land , modern mammal orders ( including primates ) suddenly appear in Europe and in North America . Sediment deposition changed significantly at many outcrops and in many drill cores spanning this time interval .   At least since 1997 , the Paleocene -- Eocene Thermal Maximum has become a focal point of considerable geoscience research because it probably provides the best past analog by which to understand impacts of global climate warming and of massive carbon input to the ocean and atmosphere , including ocean acidification . Although it is now widely accepted that the PETM represents a `` case study '' for global warming and massive carbon input to Earth 's surface , the cause , details and overall significance of the event remain perplexing .\",\n",
              "  8.232578740191368),\n",
              " (\"Hockey stick graphs present the global or hemispherical mean temperature record of the past 500 to 2000 years as shown by quantitative climate reconstructions based on climate proxy records . These reconstructions have consistently shown a slow long term cooling trend changing into relatively rapid warming in the 20th century , with the instrumental temperature record by 2000 exceeding earlier temperatures .   The term `` hockey stick graph '' was coined by the climatologist Jerry Mahlman , to describe the pattern shown by the ( MBH99 ) reconstruction , envisaging a graph that is relatively flat with a downward trend to 1900 as forming an ice hockey stick 's `` shaft '' followed by a sharp , steady increase corresponding to the `` blade '' portion . The reconstructions have featured in Intergovernmental Panel on Climate Change ( IPCC ) reports as evidence of global warming . Arguments over the reconstructions have been taken up by fossil fuel industry funded lobbying groups attempting to cast doubt on climate science .   Paleoclimatology dates back to the 19th century , and the concept of examining varves in lake beds and tree rings to track local climatic changes was suggested in the 1930s . In the 1960s , Hubert Lamb generalised from historical documents and temperature records of central England to propose a Medieval Warm Period from around 900 to 1300 , followed by Little Ice Age . This was the basis of a `` schematic diagram '' featured in the IPCC First Assessment Report of 1990 beside cautions that the medieval warming might not have been global . The use of indicators to get quantitative estimates of the temperature record of past centuries was developed , and by the late 1990s a number of competing teams of climatologists found indications that recent warming was exceptional . introduced the `` Composite Plus Scaling '' ( CPS ) method used by most later large scale reconstructions . Their study was featured in the IPCC Second Assessment Report of 1995 .   In 1998 Michael E. Mann , Raymond S. Bradley and Malcolm K. Hughes developed new statistical techniques to produce ( MBH98 ) , the first eigenvector-based climate field reconstruction ( CFR ) . This showed global patterns of annual surface temperature , and included a graph of average hemispheric temperatures back to 1400 with shading emphasising that uncertainties ( to two standard error limits ) were much greater in earlier centuries . independently produced a CPS reconstruction extending back for a thousand years , and ( MBH99 ) used the MBH98 methodology to extend their study back to 1000 .   A version of the MBH99 graph was featured prominently in the 2001 IPCC Third Assessment Report ( TAR ) , which also drew on Jones et al. 1998 and three other reconstructions to support the conclusion that , in the Northern Hemisphere , the 1990s was likely to have been the warmest decade and 1998 the warmest year during the past 1,000 years . The graph became a focus of dispute for those opposed to the strengthening scientific consensus that late 20th century warmth was exceptional . In 2003 , as lobbying over the 1997 Kyoto Protocol intensified , a paper claiming greater medieval warmth was quickly dismissed by scientists in the Soon and Baliunas controversy . Later in 2003 , Stephen McIntyre and Ross McKitrick published disputing the data used in MBH98 paper . In 2004 Hans von Storch published criticism of the statistical techniques as tending to underplay variations in earlier parts of the graph , though this was disputed and he later accepted that the effect was very small . In 2005 McIntyre and McKitrick published criticisms of the principal components analysis methodology as used in MBH98 and MBH99 . Their analysis was subsequently disputed by published papers including and which pointed to errors in the McIntyre and McKitrick methodology . Political disputes led to the formation of a panel of scientists convened by the National Research Council , their North Report in 2006 supported Mann 's findings with some qualifications , including agreeing that there were some statistical failings but these had little effect on the result .   More than two dozen reconstructions , using various statistical methods and combinations of proxy records , support the broad consensus shown in the original 1998 hockey-stick graph , with variations in how flat the pre-20th century `` shaft '' appears . The 2007 IPCC Fourth Assessment Report cited 14 reconstructions , 10 of which covered 1,000 years or longer , to support its strengthened conclusion that it was likely that Northern Hemisphere temperatures during the 20th century were the highest in at least the past 1,300 years . Further reconstructions , including Mann et al. 2008 and , have supported these general conclusions .\",\n",
              "  8.219313534493846)]"
            ]
          },
          "metadata": {},
          "execution_count": 72
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Scifact test"
      ],
      "metadata": {
        "id": "1YxFIFKTBNLM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dataset = ir_datasets.load(\"beir/scifact/train\")"
      ],
      "metadata": {
        "id": "i1g2Gf2LjToC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%capture\n",
        "for doc in dataset.docs_iter()[:1]:\n",
        "    print(doc)"
      ],
      "metadata": {
        "id": "WIDi1JwIjh7R"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dataset.docs_cls().__annotations__"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "exkc-XHOkALZ",
        "outputId": "cebc77d1-5f56-486d-8944-d2bccde462e4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "OrderedDict([('doc_id', str), ('text', str), ('title', str)])"
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "corpus=[]\n",
        "for doc in dataset.docs_iter():\n",
        "    corpus.append([doc[0],doc[1],doc[2]])\n",
        "print(len(corpus))\n",
        "documents = [doc[1] for doc in corpus]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pkhFTCerkCNh",
        "outputId": "6477bdd7-4f8d-4bf8-e137-e1c7c03f5b51"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "5183\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "vectorizer = CountVectorizer(stop_words='english', strip_accents='ascii')\n",
        "documents_vectorized = vectorizer.fit_transform(documents)\n",
        "vocabulary = vectorizer.get_feature_names_out()\n",
        "dataframe = pd.DataFrame(documents_vectorized.toarray(), columns=vocabulary)\n",
        "dataframe"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "G3E7sQ5_kg9x",
        "outputId": "36e59d9b-8be8-4246-c317-3026bffa83d3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      00  000  0000  00001  00002  000035  00004  00005  00006  00008  ...  \\\n",
              "0      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "1      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "2      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "3      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "4      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "...   ..  ...   ...    ...    ...     ...    ...    ...    ...    ...  ...   \n",
              "5178   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5179   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5180   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5181   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5182   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "\n",
              "      zygosity  zygote  zygotes  zygotic  zymography  zymosan  zyxin  zz  zzw  \\\n",
              "0            0       0        0        0           0        0      0   0    0   \n",
              "1            0       0        0        0           0        0      0   0    0   \n",
              "2            0       0        0        0           0        0      0   0    0   \n",
              "3            0       0        0        0           0        0      0   0    0   \n",
              "4            0       0        0        0           0        0      0   0    0   \n",
              "...        ...     ...      ...      ...         ...      ...    ...  ..  ...   \n",
              "5178         0       0        0        0           0        0      0   0    0   \n",
              "5179         0       0        0        0           0        0      0   0    0   \n",
              "5180         0       0        0        0           0        0      0   0    0   \n",
              "5181         0       0        0        0           0        0      0   0    0   \n",
              "5182         0       0        0        0           0        0      0   0    0   \n",
              "\n",
              "      zzz  \n",
              "0       0  \n",
              "1       0  \n",
              "2       0  \n",
              "3       0  \n",
              "4       0  \n",
              "...   ...  \n",
              "5178    0  \n",
              "5179    0  \n",
              "5180    0  \n",
              "5181    0  \n",
              "5182    0  \n",
              "\n",
              "[5183 rows x 35568 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-53ac72f5-24bc-4ef9-b057-c6c584d41614\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>00</th>\n",
              "      <th>000</th>\n",
              "      <th>0000</th>\n",
              "      <th>00001</th>\n",
              "      <th>00002</th>\n",
              "      <th>000035</th>\n",
              "      <th>00004</th>\n",
              "      <th>00005</th>\n",
              "      <th>00006</th>\n",
              "      <th>00008</th>\n",
              "      <th>...</th>\n",
              "      <th>zygosity</th>\n",
              "      <th>zygote</th>\n",
              "      <th>zygotes</th>\n",
              "      <th>zygotic</th>\n",
              "      <th>zymography</th>\n",
              "      <th>zymosan</th>\n",
              "      <th>zyxin</th>\n",
              "      <th>zz</th>\n",
              "      <th>zzw</th>\n",
              "      <th>zzz</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5178</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5179</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5180</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5181</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5182</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5183 rows × 35568 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-53ac72f5-24bc-4ef9-b057-c6c584d41614')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-53ac72f5-24bc-4ef9-b057-c6c584d41614 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-53ac72f5-24bc-4ef9-b057-c6c584d41614');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "dfs = (dataframe > 0).sum(axis=0)\n",
        "N = dataframe.shape[0]\n",
        "idfs = np.log10(N/dfs)\n",
        "\n",
        "# Lets first define all the variables we need:\n",
        "k_1 = 1.2 # single value\n",
        "b = 0.8 # single value\n",
        "## considering all words in doc\n",
        "dls = [len(d.split(' ')) for d in documents] # vector\n",
        "## considering words minus stop words in doc. \n",
        "dls = dataframe.sum(axis=1).tolist()\n",
        "avgdl = np.mean(dls) # single value\n",
        "\n",
        "numerator = np.array((k_1 + 1) * dataframe)\n",
        "denominator = np.array(k_1 *((1 - b) + b * (dls / avgdl))).reshape(N,1) + np.array(dataframe)\n",
        "BM25_tf = numerator / denominator\n",
        "idfs = np.array(idfs)\n",
        "BM25_score = BM25_tf * idfs\n",
        "\n",
        "bm25_idf = pd.DataFrame(BM25_score, columns=vocabulary)\n",
        "bm25_idf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "OtpbYAsEk7up",
        "outputId": "863231b7-9835-4224-b7da-0640a1475fd6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       00  000  0000  00001  00002  000035  00004  00005  00006  00008  ...  \\\n",
              "0     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "1     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "2     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "3     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "4     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "...   ...  ...   ...    ...    ...     ...    ...    ...    ...    ...  ...   \n",
              "5178  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5179  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5180  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5181  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5182  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "\n",
              "      zygosity  zygote  zygotes  zygotic  zymography  zymosan  zyxin   zz  \\\n",
              "0          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "1          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "2          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "3          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "4          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "...        ...     ...      ...      ...         ...      ...    ...  ...   \n",
              "5178       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5179       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5180       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5181       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5182       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "\n",
              "      zzw  zzz  \n",
              "0     0.0  0.0  \n",
              "1     0.0  0.0  \n",
              "2     0.0  0.0  \n",
              "3     0.0  0.0  \n",
              "4     0.0  0.0  \n",
              "...   ...  ...  \n",
              "5178  0.0  0.0  \n",
              "5179  0.0  0.0  \n",
              "5180  0.0  0.0  \n",
              "5181  0.0  0.0  \n",
              "5182  0.0  0.0  \n",
              "\n",
              "[5183 rows x 35568 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-955060fd-f4da-41d6-933f-9f270ff7f4e9\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>00</th>\n",
              "      <th>000</th>\n",
              "      <th>0000</th>\n",
              "      <th>00001</th>\n",
              "      <th>00002</th>\n",
              "      <th>000035</th>\n",
              "      <th>00004</th>\n",
              "      <th>00005</th>\n",
              "      <th>00006</th>\n",
              "      <th>00008</th>\n",
              "      <th>...</th>\n",
              "      <th>zygosity</th>\n",
              "      <th>zygote</th>\n",
              "      <th>zygotes</th>\n",
              "      <th>zygotic</th>\n",
              "      <th>zymography</th>\n",
              "      <th>zymosan</th>\n",
              "      <th>zyxin</th>\n",
              "      <th>zz</th>\n",
              "      <th>zzw</th>\n",
              "      <th>zzz</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5178</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5179</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5180</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5181</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5182</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5183 rows × 35568 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-955060fd-f4da-41d6-933f-9f270ff7f4e9')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-955060fd-f4da-41d6-933f-9f270ff7f4e9 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-955060fd-f4da-41d6-933f-9f270ff7f4e9');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "query_num = 1\n",
        "queries = []\n",
        "for q in dataset.queries_iter():\n",
        "    queries.append([q[0],q[1]])\n",
        "print([queries[query_num][1]])\n",
        "\n",
        "vectorizer_q = CountVectorizer(stop_words='english', strip_accents='ascii')\n",
        "query_vectorized = vectorizer_q.fit_transform([queries[query_num][1]])\n",
        "q_terms = vectorizer_q.get_feature_names_out()\n",
        "q_terms"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PotSpiIxltBy",
        "outputId": "14ca072d-603c-4a22-9de2-591acad3f63d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['1 in 5 million in UK have abnormal PrP positivity.']\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['abnormal', 'million', 'positivity', 'prp', 'uk'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "q_terms = [term for term in q_terms if term in bm25_idf.columns]\n",
        "q_terms_only_df = bm25_idf[q_terms]\n",
        "score_q_d = q_terms_only_df.sum(axis=1)\n",
        "documents = [doc[1] for doc in corpus]\n",
        "sorted(zip(documents,score_q_d.values), key = lambda tup:tup[1], reverse=True)[:10]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ypvq9CGBl6Is",
        "outputId": "635e8198-8bd0-4bab-f048-f08ea746e24c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[('OBJECTIVES To carry out a further survey of archived appendix samples to understand better the differences between existing estimates of the prevalence of subclinical infection with prions after the bovine spongiform encephalopathy epizootic and to see whether a broader birth cohort was affected, and to understand better the implications for the management of blood and blood products and for the handling of surgical instruments. DESIGN Irreversibly unlinked and anonymised large scale survey of archived appendix samples. SETTING Archived appendix samples from the pathology departments of 41 UK hospitals participating in the earlier survey, and additional hospitals in regions with lower levels of participation in that survey. SAMPLE 32,441 archived appendix samples fixed in formalin and embedded in paraffin and tested for the presence of abnormal prion protein (PrP). RESULTS Of the 32,441 appendix samples 16 were positive for abnormal PrP, indicating an overall prevalence of 493 per million population (95% confidence interval 282 to 801 per million). The prevalence in those born in 1941-60 (733 per million, 269 to 1596 per million) did not differ significantly from those born between 1961 and 1985 (412 per million, 198 to 758 per million) and was similar in both sexes and across the three broad geographical areas sampled. Genetic testing of the positive specimens for the genotype at PRNP codon 129 revealed a high proportion that were valine homozygous compared with the frequency in the normal population, and in stark contrast with confirmed clinical cases of vCJD, all of which were methionine homozygous at PRNP codon 129. CONCLUSIONS This study corroborates previous studies and suggests a high prevalence of infection with abnormal PrP, indicating vCJD carrier status in the population compared with the 177 vCJD cases to date. These findings have important implications for the management of blood and blood products and for the handling of surgical instruments.',\n",
              "  11.537320382799532),\n",
              " ('We report a case of preclinical variant Creutzfeldt-Jakob disease (vCJD) in a patient who died from a non-neurological disorder 5 years after receiving a blood transfusion from a donor who subsequently developed vCJD. Protease-resistant prion protein (PrP(res)) was detected by western blot, paraffin-embedded tissue blot, and immunohistochemistry in the spleen, but not in the brain. Immunohistochemistry for prion protein was also positive in a cervical lymph node. The patient was a heterozygote at codon 129 of PRNP, suggesting that susceptibility to vCJD infection is not confined to the methionine homozygous PRNP genotype. These findings have major implications for future estimates and surveillance of vCJD in the UK.',\n",
              "  6.3384804894944935),\n",
              " (\"The prion protein (PrP(C)) is highly expressed in the nervous system and critically involved in prion diseases where it misfolds into pathogenic PrP(Sc). Moreover, it has been suggested as a receptor mediating neurotoxicity in common neurodegenerative proteinopathies such as Alzheimer's disease. PrP(C) is shed at the plasma membrane by the metalloprotease ADAM10, yet the impact of this on prion disease remains enigmatic. Employing conditional knockout mice, we show that depletion of ADAM10 in forebrain neurons leads to posttranslational increase of PrP(C) levels. Upon prion infection of these mice, clinical, biochemical, and morphological data reveal that lack of ADAM10 significantly reduces incubation times and increases PrP(Sc) formation. In contrast, spatiotemporal analysis indicates that absence of shedding impairs spread of prion pathology. Our data support a dual role for ADAM10-mediated shedding and highlight the role of proteolytic processing in prion disease.\",\n",
              "  5.479213806892906),\n",
              " ('Native mammalian prions exist in self-propagating strains that exhibit distinctive clinical, pathological and biochemical characteristics. Prion strain diversity is associated with variations in PrP(Sc) conformation, but it remains unknown precisely which physical properties of the PrP(Sc) molecules are required to encipher mammalian prion strain phenotypes. In this study, we subjected prion-infected brain homogenates derived from three different hamster scrapie strains to either (i) proteinase K digestion or (ii) sonication, and inoculated the modified samples into normal hamsters. The results show that the strain-specific clinical features and neuropathological profiles of inoculated animals were not affected by either treatment. Similarly, the strain-dependent biochemical characteristics of the PrP(Sc) molecules (including electrophoretic mobility, glycoform composition, conformational stability and susceptibility to protease digestion) in infected animals were unaffected by either proteolysis or sonication of the original inocula. These results indicate that the infectious strain properties of native prions do not appear to be altered by PrP(Sc) disaggregation, and that maintenance of such properties does not require the N-domain (approximately residues 23-90) of the protease-resistant PrP(Sc) molecules or protease-sensitive PrP(Sc) molecules.',\n",
              "  5.42241988394522),\n",
              " ('Prions are the protein-based infectious agents responsible for prion diseases. Environmental prion contamination has been implicated in disease transmission. Here, we analyzed the binding and retention of infectious prion protein (PrP(Sc)) to plants. Small quantities of PrP(Sc) contained in diluted brain homogenate or in excretory materials (urine and feces) can bind to wheat grass roots and leaves. Wild-type hamsters were efficiently infected by ingestion of prion-contaminated plants. The prion-plant interaction occurs with prions from diverse origins, including chronic wasting disease. Furthermore, leaves contaminated by spraying with a prion-containing preparation retained PrP(Sc) for several weeks in the living plant. Finally, plants can uptake prions from contaminated soil and transport them to aerial parts of the plant (stem and leaves). These findings demonstrate that plants can efficiently bind infectious prions and act as carriers of infectivity, suggesting a possible role of environmental prion contamination in the horizontal transmission of the disease.',\n",
              "  4.827903110869903),\n",
              " ('OBJECTIVE To assess the robustness of patient responses to a new national survey of patient experience as a basis for providing financial incentives to doctors. DESIGN Analysis of the representativeness of the respondents to the GP Patient Survey compared with those who were sampled (5.5 million patients registered with 8273 general practices in England in January 2009) and with the general population. Analysis of non-response bias looked at the relation between practice response rates and scores on the survey. Analysis of the reliability of the survey estimated the proportion of the variance of practice scores attributable to true differences between practices. RESULTS The overall response rate was 38.2% (2.2 million responses), which is comparable to that in surveys using similar methodology in the UK. Men, young adults, and people living in deprived areas were under-represented among respondents. However, for questions related to pay for performance, there was no systematic association between response rates and questionnaire scores. Two questions which triggered payments to general practitioners were reliable measures of practice performance, with average practice-level reliability coefficients of 93.2% and 95.0%. Less than 3% and 0.5% of practices had fewer than the number of responses required to achieve conventional reliability levels of 90% and 70%. A change to the payment formula in 2009 resulted in an increase in the average impact of random variation in patient scores on payments to general practitioners compared with payments made in 2007 and 2008. CONCLUSIONS There is little evidence to support the concern of some general practitioners that low response rates and selective non-response bias have led to systematic unfairness in payments attached to questionnaire scores. The study raises issues relating to the validity and reliability of payments based on patient surveys and provides lessons for the UK and for other countries considering the use of patient experience as part of pay for performance schemes.',\n",
              "  4.802553037383037),\n",
              " ('Infectious prions propagate from peripheral entry sites into the central nervous system (CNS), where they cause progressive neurodegeneration that ultimately leads to death. Yet the pathogenesis of prion disease can vary dramatically depending on the strain, or conformational variant of the aberrantly folded and aggregated protein, PrP(Sc). Although most prion strains invade the CNS, some prion strains cannot gain entry and do not cause clinical signs of disease. The conformational basis for this remarkable variation in the pathogenesis among strains is unclear. Using mouse-adapted prion strains, here we show that highly neuroinvasive prion strains primarily form diffuse aggregates in brain and are noncongophilic, conformationally unstable in denaturing conditions, and lead to rapidly lethal disease. These neuroinvasive strains efficiently generate PrP(Sc) over short incubation periods. In contrast, the weakly neuroinvasive prion strains form large fibrillary plaques and are stable, congophilic, and inefficiently generate PrP(Sc) over long incubation periods. Overall, these results indicate that the most neuroinvasive prion strains are also the least stable, and support the concept that the efficient replication and unstable nature of the most rapidly converting prions may be a feature linked to their efficient spread into the CNS.',\n",
              "  4.688771720724594),\n",
              " ('Hemolysates of erythrocytes from more than a quarter million people in Alabama were electrophoresed on cellulose acetate, pH 8.4, and those samples exhibiting an abnormality were also electrophoresed in citrate agar, pH 6.0. The globin chains of mutants other than Hb S and C were electrophoresed in urea-mercaptoethanol buffers at both pH 8.9 and pH 6.0, and 60 of them were also analyzed structurally. Of about 6000 samples from whites, only three contained abnormal hemoglobins--Hb D Los Angeles, Hb J Baltimore, and one unidentified. Of 249,000 samples from blacks, about 29,000 contained electrophoretically detectable abnormalities, most of them associated with Hb S or C, present in a frequency of about 9% and 3%, respectively. About 1000 samples resolved into patterns of potential clinical significance. Twenty other mutant hemoglobins were detected, in various genetic combinations in 164 kindreds; four of these-Hb Alabama, Montgomery, Titusville, and Mobile--were previously unknown. The methods used are rapid, economical, and well suited for large scale surveys. They provide highly specific characterizations of many mutant hemoglobins, and no discrepancies were found between the presumptive identifications based on these characterizations and the definitive identifications obtained from structural analyses.',\n",
              "  3.8640302304421676),\n",
              " ('A considerable subgroup of patients with early breast cancer does not address benefits of anthracycline based chemotherapy. The aim of this retrospective study was to investigate the effect of microvessel density (MVD) and status of p53 protein on 5-year disease free survival (DFS) in the group of breast cancer patients treated with anthracyclines in adjuvant setting. Correlations between MVD, p53 status and other clinicopathological parameters were also assessed. MVD and p53 status were analyzed immunohistochemically in the group of 172 women with breast cancer in clinical stage T1-2, N1-N2, M0. There were 123 tumors (71.5 %) with lower MVD (≤ 214.8 microvesells/mm(2)) and 49 (28.5 %) with higher MVD (>214.8 microvesells/mm(2)). The proportion of higher MVD tumors significantly increased in N2 (P = 0.000) and in estrogen (P = 0.046) or progesterone receptors (P = 0.029) negative tumors. p53 positivity was indicated in 50 cancers (29.1 %) and was significantly associated with higher grade (P = 0.000), steroid receptors negativity (P = 0.000), cytokeratin5/6 positivity (P = 0.026), topoisomerase IIα overexpression (P = 0.005) and higher proliferation rate (P = 0.001). In univariate analysis, higher MVD (P = 0.016) and p53 negativity (P = 0.023) were significantly related with longer DFS (median follow-up 36 months). In multivariate Cox regression analysis MVD was independently associated with DFS. These data suggest that higher MVD is favourable prognostic factors for early advanced breast cancer patients after adjuvant anthracycline based chemotherapy.',\n",
              "  3.6448981560117626),\n",
              " ('OBJECTIVE To determine whether myelin oligodendrocyte glycoprotein antibodies (MOG-Abs) were predictive of a demyelination phenotype in children presenting with acquired demyelinating syndrome (ADS). METHOD Sixty-five children with a first episode of ADS (12 acute disseminated encephalomyelitis, 24 optic neuritis, 18 transverse myelitis, 11 other clinically isolated syndrome) were identified from 2 national demyelination programs in the United Kingdom and France. Acute serum samples were tested for MOG-Abs by cell-based assay. Antibodies were used to predict diagnosis of multiple sclerosis (MS) at 1 year. RESULTS Twenty-three of 65 (35%) children had MOG-Abs. Antibody-positive and antibody-negative patients were not clinically different at presentation, but identification of MOG-Abs predicted a non-MS course at 1-year follow-up: only 2/23 (9%) MOG-Ab-positive patients were diagnosed with MS compared to 16/42 (38%) MOG-Ab-negative patients (p = 0.019, Fisher exact test). Antibody positivity at outset was a useful predictor for a non-MS disease course, with a positive predictive value of 91% (95% confidence interval [CI] 72-99), negative predictive value of 38% (95% CI 24-54), positive likelihood ratio of 4.02 (CI 1.0-15.4), and odds ratio of 6.5 (CI 1.3-31.3). CONCLUSIONS MOG-Abs are found at presentation in 35% of patients with childhood ADS, across a range of demyelinating disorders. Antibody positivity can be useful in predicting a non-MS disease course at onset.',\n",
              "  3.621069192264962)]"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    }
  ]
}